### PR TITLE
Merge pirocchi/dotnet/merge-develop into develop

### PR DESCRIPTION
**Pull request by mpiroc**
## Description

Merge the feature branch pirocchi/dotnet/merge-develop back into develop.

## Motivation and Context

The feature branch is complete.

## Testing

1. Created a temporary branch off of `develop`, and merge `merge-develop` into that branch.
2. Run all tests from temporary branch.
3. Performed sanity-level manual testing on Windows from temporary branch.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

Note: I will add a CHANGELOG entry and update this PR.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -82,6 +83,9 @@ export async function activate(context: vscode.ExtensionContext) {
             }
         }
 
+        ext.awsContextCommands = new DefaultAWSContextCommands(awsContext, awsContextTrees, regionProvider)
```
Comment (awschristou) [inline src/extension.ts]: suspect diff - see lines 71-73

File: src/extension.ts [hunk 2]
```diff
@@ -103,7 +107,7 @@ export async function activate(context: vscode.ExtensionContext) {
             callback: async () => await ext.awsContextCommands.onCommandLogin(),
             telemetryName: {
                 namespace: TelemetryNamespace.Aws,
-                name: 'credentialslogin'
+                name: 'credentialsLogin'
```
Comment (awschristou) [inline src/extension.ts]: @bryceitoc9 can you confirm the proper name of this metric?

Comment (bryceitoc9) [reply src/extension.ts]: Confirmed all lowercase is correct.

File: src/lambda/models/samLambdaRuntime.ts [hunk 3]
```diff
@@ -13,50 +13,25 @@ export type SamLambdaRuntime =
...
-    'dotnetcore1.0' |
-    'dotnetcore' |
-    'dotnet' |
-    'go1.x' |
-    'go' |
-    'java8' |
-    'java' |
-    'ruby' |
-    'ruby2.5'
+    'dotnetcore2.0'
```
Comment (bryceitoc9) [inline src/lambda/models/samLambdaRuntime.ts]: Should this be removed? You can't deploy new `dotnetcore2.0` apps.

File: src/lambda/models/samLambdaRuntime.ts [hunk 4]
```diff
@@ -13,50 +13,25 @@ export type SamLambdaRuntime =
...
-    'dotnetcore1.0',
-    'dotnetcore',
-    'dotnet',
-    'go1.x',
-    'go',
-    'java8',
-    'java',
-    'ruby',
-    'ruby2.5'
+    'dotnetcore2.0'
```
Comment (bryceitoc9) [inline src/lambda/models/samLambdaRuntime.ts]: Ditto above

File: src/shared/codelens/codeLensUtils.ts [hunk 5]
```diff
@@ -20,7 +20,7 @@ import { defaultMetricDatum } from '../telemetry/telemetryUtils'
 import { toArrayAsync } from '../utilities/collectionUtils'
 import { localize } from '../utilities/vsCodeUtils'
 
-export type Language = 'python' | 'javascript'
+export type Language = 'python' | 'javascript' | 'csharp'
```
Comment (bryceitoc9) [inline src/shared/codelens/codeLensUtils.ts]: Nit: should we rename `typescriptCodeLensProvider` to `javascriptCodeLensProvider`?

Comment (awschristou) [reply src/shared/codelens/codeLensUtils.ts]: Out of scope for the feature branch merge

**Comment by mpiroc** [comment]
Due to a series of misteps, we are squashing this entire feature branch into develop.

1. We allowed `feature/dotnet` to drift too far from `develop`, which required significant manual conflict resolution to resolve.
2. We should have squashed the conflict-resolution history *before* merging develop, but we did not. Therefore in order to squash the pre-merge commits, we would need to recreate the merge and re-resolve conflicts. This would reduce our confidence in the quality of the branch.

In the future:
1. We will regularly merge into feature branches to prevent drift.
2. If we do encounter a situation that involves manually porting, squash the ports *before* merging develop.

**Pull request merged**